security/src/main/scala/pl/touk/nussknacker/ui/security/api/LoggedUser.scala

```diff
@@ -11,12 +13,33 @@ sealed trait LoggedUser {
   def can(category: String, permission: Permission): Boolean
 }
 
+sealed trait RealLoggedUser extends LoggedUser
+
 object LoggedUser {
 
-  def apply(
+  def create(
       authenticatedUser: AuthenticatedUser,
       rules: List[ConfigRule]
-  ): LoggedUser = {
+  ): Either[CreationError, LoggedUser] = {
+    val loggedUser = LoggedUser(authenticatedUser, rules)
+    authenticatedUser.impersonatedAuthenticationUser match {
```

Comment:
we can flatten it a little bit:
```scala
    authenticatedUser.impersonatedAuthenticationUser match {
      case None =>
        Right(loggedUser)
      case Some(impersonatedUser) if loggedUser.canImpersonate =>
        Right(
          ImpersonatedUser(
            impersonatedUser = LoggedUser(impersonatedUser, rules),
            impersonatingUser = loggedUser
          )
        )
      case Some(_) =>
        Left(ImpersonationNotAllowed)
    }
```

WDYT?

Reply:
Nice! I modified it to be this way, thank you!